packages/relay/src/lib/eth.ts

```diff
@@ -1320,24 +1320,31 @@ export class EthImpl implements Eth {
       return [];
     }
 
-    return result.logs.map(log => {
-      return new Log({
-        address: log.address,
-        blockHash: EthImpl.toHash32(log.block_hash),
-        blockNumber: EthImpl.numberTo0x(log.block_number),
-        data: log.data,
-        logIndex: EthImpl.numberTo0x(log.index),
-        removed: false,
-        topics: log.topics,
-        transactionHash: EthImpl.toHash32(log.transaction_hash),
-        transactionIndex: EthImpl.numberTo0x(log.transaction_index)
-      });
-    });
+    return Promise.all(
+      result.logs.map(async log => {
+        return new Log({
+          address: await this.getContractEvmAddress(log.address, requestId),
```

Comment:
The log address can repeat multiple times within the payload. `/api/v1/contracts/{id}` is a heavyweight call that returns large byte arrays. We should minimize calls to this API by collecting `log.address` into a set then calling the distinct addresses outside this loop. It's a performance optimization that makes the code not as clean but is necessary to reduce the load on mirror node.

Reply:
Agreed.
Prefer the caching option since this is just a workaround till Mirror Node adopts similar logic.
Then this work would go away.